### PR TITLE
Remove `dev/linux`, `dev/macos`, `dev/windows` in favor of streamlined GH workflow

### PR DESCRIPTION
**Pull request by kenodegard**
### Description

Changes include:
- Bring `dev/linux` inline in `tests.yml`
- Remove linux images
- Use standard QEMU images instead of custom images
- Bring `dev/macos` inline in `tests.yml`
- Bring `dev/windows` inline in `tests.yml`
- ~Enable alpha sorting on `requirements.txt` files~ (cherry picked into https://github.com/conda/conda/pull/13166) 

Updates to `tests.yml`:
- Test matrix runs 3 integration tests and 2 unit tests per OS/Python/channel combo
- ~Adds Allure reports~ (no longer desired)

Two bash activation tests on Windows proved difficult to resolve neatly within this already large PR. Opted to mark them as skip and to tackle them separately.

Xref https://github.com/conda/conda-build/pull/5019
Resolves https://github.com/conda/conda/issues/12418

### Checklist - did you ...

- [ ] ~Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?~
     
- [ ] ~Add / update necessary tests?~
- [ ] ~Add / update outdated documentation?~

### CODE REVIEW

**Comment by jezdez** [comment]
@kenodegard FWIW conda-libmamba-solver is using the [images to run conda tests](https://github.com/conda/conda-libmamba-solver/blob/17c8338029b1c951da2c34d4a8e4dfc5be7e69b3/.github/workflows/tests.yml#L65) and it's super critical to not do a pull rug just before we do the October release.

There is probably a great opportunity to make it easier to maintain the cls tests as well, just be aware that it does take additional effort to stabilize such a change, please include that in your risk assessment. I'd love to get you and Jaime coordinate on this, to decide in which order and timeframe this change may be applied in cls as well.

**Comment by jezdez** [comment]
@kenodegard Now that we're depending on conda-libmamba-solver, I'm wondering if it's time to reinvestigate this, since we may want to reduce the duplication of test runs between the conda repo and the conda-libmamba-solver repo. @jaimergp any suggestion related to this?

File: .github/workflows/tests.yml
Comment (jezdez) [inline .github/workflows/tests.yml:82]: Does this mean "automatic"?

File: .github/workflows/tests.yml
Comment (jezdez) [inline .github/workflows/tests.yml:81]: Oooh, that's smart! Much nicer :)

File: .github/workflows/tests.yml
Comment (jezdez) [inline .github/workflows/tests.yml:106]: This'll run the tests in a non-base environment, are we sure that this won't be a problem with tests that assume to be running in base?

File: .github/workflows/tests.yml
Comment (jezdez) [inline .github/workflows/tests.yml:319]: We could combine this with the benchmark tests instead of this arbitrary old test?

**Review by jezdez — COMMENTED** [review]
This is an amazing attempt to clean up some of this mess, but there are open questions:

- Without the Docker image, we should remove the gitpod config and docs
- What's the expected developer experience for dev/CI parity? Can we use https://github.com/nektos/act to run the same tests locally (at least on the native platforms)?
- How can conda-libmamba-solver run the conda tests with this new regime without access to pre-built Docker images?

File: conda/core/package_cache_data.py
Comment (jaimergp) [inline conda/core/package_cache_data.py:78]: I'm still a bit green with `:=` does this assignment propagate to the block under `else` or is it local to this block only?

File: tests/cli/test_main_clean.py
Comment (jaimergp) [inline tests/cli/test_main_clean.py:82]: can we assert this here just in case? It was the original issue:

```
assert context.pkgs_dirs == (tmp_pkgs_dir,)
```

File: conda/core/package_cache_data.py
Comment (kenodegard) [reply conda/core/package_cache_data.py:78]: yes, it propagates

it's a normal local variable, so any subsequent code can access it

the walrus operator did not change how scopes work

File: tests/cli/test_main_clean.py
Comment (kenodegard) [reply tests/cli/test_main_clean.py:82]: this assertion is in the new `tmp_pkgs_dir` fixture

Comment (jaimergp) [reply tests/cli/test_main_clean.py:82]: It's only part of it. It tests that it's the first one, but it doesn't assert that it's the only one:

```
assert context.pkgs_dirs[0] == pkgs_dir_str
```

In `test_main_clean` we need `context.pkgs_dirs` to ignore user config. Otherwise prepopulated caches might affect the outcome of the test (e.g. the user cache at ~/.conda/pkgs already has `zlib` (now `small-package`?) due to other tests or CI configs, so conda won't download it, so the new tmp pkgs_dir is not populated).

Comment (kenodegard) [reply tests/cli/test_main_clean.py]: opted to mock `context.pkgs_dirs` instead of monkeypatching to resolve this issue, see https://github.com/conda/conda/pull/13511

File: .github/workflows/tests.yml
Comment (kenodegard) [reply .github/workflows/tests.yml]: looks like we're good, we seem to have naturally drifted away from this expectation

File: .github/workflows/tests.yml
Comment (kenodegard) [reply .github/workflows/tests.yml]: the docs only says this is the default value if `--pdb` is specified

looking at pytest-xdist's source it looks like a zero value means it runs inline to the spawning process, with `-n 1` it will spawn a child `popen` process to run the tests

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:121]: I'd also add a `conda config --show-sources`.

**Comment by jaimergp** [comment]
> What's the expected developer experience for dev/CI parity? Can we use https://github.com/nektos/act to run the same tests locally (at least on the native platforms)?

I'm also interested about the DX story for interactive debugging locally.

> How can conda-libmamba-solver run the conda tests with this new regime without access to pre-built Docker images?

I guess our only chance is to copy things and adjust the steps, and then have an upstream checker to manually port changes in the CI files ([something similar to this kind of system](https://github.com/napari/packaging/issues/105) we have in napari/packaging). Reusable workflows are tricky because we'd need to add extra steps and modify some commands and it's more trouble than it's worth for just one project, IMO.

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:43]: The `filters` argument here needs to be updated to drop the `dev/` files.

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:1]: Would you consider adding a single, yet parametrized when appropriate, workflow job for all platforms? I see a lot of duplication, but maybe the parametrization is more trouble than it's worth.

Comment (kenodegard) [reply .github/workflows/tests.yml:1]: I was thinking to try that in a subsequent PR

File: .github/workflows/tests.yml
Comment (kenodegard) [reply .github/workflows/tests.yml:120]: https://github.com/conda/conda/pull/13162/commits/80e38e8a3ce880ffcb57ebfb9a6723a51db5752c

File: .github/workflows/tests.yml
Comment (kenodegard) [reply .github/workflows/tests.yml:43]: https://github.com/conda/conda/pull/13162/commits/e7fc559182b3c8dbd9271a664639e362f7c06ac3

File: .github/workflows/tests.yml
Comment (kenodegard) [reply .github/workflows/tests.yml]: we could but why? combining them would results in benchmarks being run on all three archs (`linux-64`, `linux-arm64`, and `linux-ppc64le`)

**Comment by kenodegard** [comment]
> > What's the expected developer experience for dev/CI parity? Can we use [nektos/act](https://github.com/nektos/act) to run the same tests locally (at least on the native platforms)?
> 
> I'm also interested about the DX story for interactive debugging locally.

I don't see any reason to change the DX, if anything this change to CI actually aligns it more closely with the local DX workflow.

> > How can conda-libmamba-solver run the conda tests with this new regime without access to pre-built Docker images?
> 
> I guess our only chance is to copy things and adjust the steps, and then have an upstream checker to manually port changes in the CI files ([something similar to this kind of system](https://github.com/napari/packaging/issues/105) we have in napari/packaging). Reusable workflows are tricky because we'd need to add extra steps and modify some commands and it's more trouble than it's worth for just one project, IMO.

Yeah I expect CLS to either completely diverge from conda's CI workflow and build their own images or adopt these proposed changes.

**Comment by kenodegard** [comment]
Hm the `windows, conda-forge, integration` tests are take significantly longer to run than `windows, defaults, integration` (we're looking at 40-45min versus 3+hrs)

**Comment by jaimergp** [comment]
> Hm the `windows, conda-forge, integration` tests are take significantly longer to run than `windows, defaults, integration` (we're looking at 40-45min versus 3+hrs)

Maybe a missing `--override-channels` in some tests. Some assume defaults as default and end up mixing channels which makes classic cry.

**Comment by jaimergp** [comment]
I'd expect failures. osx-arm64 channels are less populated and I'm sure we'll find old tests that assumed availability of some version in osx-64. That is to say that it won't be a single commit change (changing the runner label) and maybe it's better to isolate that change in a separate PR.

**Comment by kenodegard** [comment]
pytest-codspeed appears to mess with how `stdout`/`stderr` are handled and so `capsys` is unable to properly intercept the streams which is why our benchmark tests are failing, see https://github.com/CodSpeedHQ/pytest-codspeed/issues/24

**Comment by kenodegard** [comment]
We're waiting on https://github.com/CodSpeedHQ/pytest-codspeed/pull/25 so until then we are using my pytest-codspeed fork

File: tests/__init__.py
Comment (jezdez) [inline tests/__init__.py:14]: 🔥 🤗 🔥

**Review by jezdez — DISMISSED** [review]

**Review by beeankha — APPROVED** [review]

File: tests/test_create.py
Comment (jaimergp) [inline tests/test_create.py:1408]: This might add channels to the solver when  `conda-forge` is the default channel. Can we instead use a conda-forge shortcut? e.g. https://anaconda.org/conda-forge/miniforge_console_shortcut

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:205]:
```suggestion
        with:
          condarc-file: .github/condarc-${{ matrix.default-channel }}
          run-post: false  # skip post cleanup
```

These options are not used in the other CI entries. Also, since we are running `conda install` next, we can save one solve and add `python=${{ matrix.python-version }}` instead there.

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:214]:
```suggestion
          ${{ env.REQUIREMENTS_TRUSTSTORE }}
          python=${{ matrix.python-version }}
```

Save one solve

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:217]: Don't we want `conda info --all`?

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:286]:
```suggestion
```

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:296]:
```suggestion
          ${{ env.REQUIREMENTS_TRUSTSTORE }}
          python=${{ matrix.python-version }}
```

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:298]:
```suggestion
      - name: Install CodSpeed
```

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:365]:
```suggestion
          && conda info --all
```

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:357]:
```suggestion
          && set -x
          && conda create
```

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:440]:
```suggestion
          ${{ env.REQUIREMENTS_TRUSTSTORE }}
          python=${{ matrix.python-version }}
```

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:431]: This will always install Miniconda, won't it? What about:

```suggestion
          miniconda-version: ${{ matrix.arch == 'osx-arm64' && 'latest' || null }}  # Miniconda not preinstalled in arm64 runners
```

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:443]:
```suggestion
        run: conda info --all
```

**Comment by jaimergp** [comment]
It does look solver related:

```
4083.60s call     tests/cli/test_env.py::test_update[classic]
2469.53s call     tests/cli/test_env.py::test_update_env_only_pip_json_output[classic]
2288.18s call     tests/cli/test_env.py::test_create_dry_run_yaml[classic]
1761.93s call     tests/cli/test_env.py::test_create_valid_env_json_output[classic]
1760.75s call     tests/cli/test_env.py::test_remove_dry_run[classic]
1755.67s call     tests/cli/test_env.py::test_set_unset_env_vars[classic]
424.98s call     tests/test_create.py::test_shortcut_absent_when_condarc_set[classic]
306.58s call     tests/cli/test_cli_install.py::test_find_conflicts_called_once[classic]
279.13s call     tests/cli/test_main_remove.py::test_remove_globbed_package_names[classic]
261.81s call     tests/test_create.py::test_dont_remove_conda_2[classic]
120.10s call     tests/test_create.py::test_install_tarball_from_file_based_channel[classic]
```

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:486]:
```suggestion
        uses: actions/download-artifact@v4
```

To fix warnings:

> aggregate
> Node.js 16 actions are deprecated. Please update the following actions to use Node.js 20: actions/download-artifact@v3. For more information see: https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20/.

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:509]: Dependabot usually recommends pinning the hash to avoid tag replacement attacks. Dependabot can recognize the hashes as version approach and update them as necessary too, so I'd recommend switching back to hashes in 3rd party (i.e. no `actions/*`) actions.

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:493]: There's a warning now in the Summary:

> [aggregate](https://github.com/conda/conda/actions/runs/8058911732/job/22096776879#step:3:10)
No files were found with the provided path: test-results-*. No artifacts will be uploaded.

File: .github/workflows/images.yml
Comment (jaimergp) [inline .github/workflows/images.yml:1]: I think this should be announced in the `news/`. And I'd also check if the docs require updates.

Comment (jaimergp) [reply .github/workflows/images.yml:1]: Also, we should decide if this means we will stop supporting DevContainer and Gitpod, which are not removed in this PR. Maybe the can be reconfigured to use the regular Miniconda/Miniforge images, but right now they point at the "latest" CI images.

Comment (jaimergp) [reply .github/workflows/images.yml:1]: And speaking of `news/`, maybe we should also mention that this PR fixes https://github.com/conda/conda/issues/12418 as the description states.

File: .github/workflows/tests.yml
Comment (kenodegard) [reply .github/workflows/tests.yml:509]: From my experience I haven't seen dependabot updating hashes when a new version was available, perhaps theres some config option we need to enable?

File: tests/test_create.py
Comment (kenodegard) [reply tests/test_create.py]: See https://github.com/conda/conda/pull/13162/commits/d6748bb3920533320ef2a276f5c4c06f7eb79a86

File: .github/workflows/tests.yml
Comment (kenodegard) [inline .github/workflows/tests.yml:103]:
```suggestion
```

File: .github/workflows/tests.yml
Comment (kenodegard) [inline .github/workflows/tests.yml:113]:
```suggestion
          ${{ env.REQUIREMENTS_TRUSTSTORE }}
          python=${{ matrix.python-version }}
```

File: .github/workflows/tests.yml
Comment (kenodegard) [inline .github/workflows/tests.yml:429]:
```suggestion
```

File: pyproject.toml
Comment (kenodegard) [inline pyproject.toml:123]:
```suggestion
# E501 line too long
```

File: .github/workflows/tests.yml
Comment (kenodegard) [reply .github/workflows/tests.yml:355]: and we don't want to do `bash -lcx ...`?

**Comment by kenodegard** [comment]
> Something about the benchmark tests looks like mixed channels as well, running way too long.

https://docs.codspeed.io/faq#why-are-my-benchmarks-taking-more-time-when-run-with-the-codspeed-github-action

File: .github/workflows/tests.yml
Comment (dholth) [inline .github/workflows/tests.yml:411]: Is this line continuing to do fewer unit test groups on OSX to preserve limited resources?

File: .github/workflows/tests.yml
Comment (kenodegard) [reply .github/workflows/tests.yml:502]: I'm not seeing this in the latest run, can you confirm?

File: .github/workflows/tests.yml
Comment (kenodegard) [reply .github/workflows/tests.yml:411]: yup, the macOS matrix only runs 5 jobs:

- osx-64 + Python 3.8 + defaults + unit 1
- osx-64 + Python 3.8 + defaults + unit 2
- osx-arm64 + Python 3.12 + conda-forge + integration 1
- osx-arm64 + Python 3.12 + conda-forge + integration 2
- osx-arm64 + Python 3.12 + conda-forge + integration 3

File: .github/workflows/images.yml
Comment (kenodegard) [reply .github/workflows/images.yml:1]: See https://github.com/conda/conda/pull/13162/commits/e6fd209145979c755b2a761068fa91b3684151ef

**Review by jezdez — APPROVED** [review]
A nice to have would be to move the codspeed action inline while we're waiting for the PR to merge upstream, or add it to conda/actions. I don't mind either way, and we can also do it as a post-merge step

**Review by beeankha — APPROVED** [review]

**Review by dbast — APPROVED** [review]

File: .github/workflows/tests.yml
Comment (jaimergp) [inline .github/workflows/tests.yml:440]: This construct might be useful in the other OSs as well, to get a fresh conda-forge-provided install with Miniforge, which might help with the long integration runtimes on Linux & Windows items that test against conda-forge.

File: .github/workflows/tests.yml
Comment (jaimergp) [reply .github/workflows/tests.yml:503]: Confirmed!

File: .github/workflows/tests.yml
Comment (jaimergp) [reply .github/workflows/tests.yml:519]: See this one PR https://github.com/conda-forge/docker-images/pull/256. [Config here](https://github.com/conda-forge/docker-images/blob/main/.github/dependabot.yml).

Comment (kenodegard) [reply .github/workflows/tests.yml:519]: ah good good, I've opened https://github.com/conda/conda/issues/13657 to address this

**Pull request merged**
